sqlmesh/core/plan/definition.py

```diff
@@ -238,13 +227,48 @@ def missing_intervals(self) -> t.List[SnapshotIntervals]:
 
     @property
     def snapshots(self) -> t.List[Snapshot]:
-        """Gets all the snapshots in the plan/environment."""
-        return list(self.context_diff.snapshots.values())
+        """Gets all the valid snapshots in the plan/environment."""
+        return [
+            snapshot
+            for snapshot in [
+                self._get_snapshot(snapshot_name)
+                for snapshot_name in self._dag
+                if snapshot_name not in self.invalid_snapshot_names
+            ]
+            if snapshot
+        ]
 
     @property
     def new_snapshots(self) -> t.List[Snapshot]:
         """Gets only new snapshots in the plan/environment."""
-        return list(self.context_diff.new_snapshots.values())
+        return [
+            snapshot
+            for snapshot in self.snapshots
+            if snapshot.snapshot_id in self.context_diff.new_snapshots
+        ]
+
+    @property
+    def has_changes(self) -> bool:
```

Comment:
Seems a bit confusing that plan and context_diff have different notions of whether there are changes or not. Can we consolidate? Eg. can invalid snapshot names be moved into ContextDiff?